### PR TITLE
feat(db_engine_specs): big query cost estimation

### PR DESCRIPTION
**Pull request by zamar-roura**
### SUMMARY
We are adding four methods inside superset/db_engine_specs/bigquery.py

For now only Postgres and Presto had cost estimation capabilities. For that they used the **cursor** object. 
Bigquery needs to use a **dry run** to get the query cost estimation. Sadly, you can't use the cursor to execute dry runs (I have tried). And that's why there is a need to override the two functions. **estimate_statement_cost** and **estimate_query_cost**. The other two methods **get_allow_cost_estimate**  and **query_cost_formatter** are just added because the base class doesn't have functionality or gives a False. 

Also the QueryEditor.py object doesn't have sqlEditorId anymore in it's variables and the Cost Estimation Reducer is looking for it, thus giving error, that is also changed. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
BEFORE 

image
 
After
image

### TESTING INSTRUCTIONS
Add ESTIMATE_QUERY_COST = True inside superset/config.py.
Add a valid BigQuery database, activate option *Enable query cost estimation* inside advanced options.
Go to SQL Lab and put a valid query, then press Estimate Cost button.

### ADDITIONAL INFORMATION
- [ X] Has associated issue: https://github.com/apache/superset/issues/20832
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ X] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by zamar-roura** [comment]
Thanks for running the tests.

I'll check on the lint errors and try to see why the cypress test is showing the "cant create chart from query error. Maybe the estimate cost button is changing the selector of the other button? Lets see.

**Comment by zamar-roura** [comment]
@villebro is it okay like this? The cypress E2E tests sometimes work sometimes don't, but everything else is fixed and working.

**Comment by zamar-roura** [comment]
Do I need to do something else to merge this?

**Comment by fernandofernandezgonzalez** [comment]
Wowwww!!! This is a so, so much needed feature!!

**Comment by villebro** [comment]
@zamar-roura would you be able to rebase the PR to resolve the conflicts?

**Comment by zamar-roura** [comment]
@villebro Tell me if you need anything else. I'll be 

The reducer/SqlLab.js needed some fixes in the ESTIMATE QUERY action as action.query didn't have any sqlEditorId parameter. It always comes as action.query.id not action.query.sqlEditorId.

**Comment by villebro** [comment]
Thanks @zamar-roura - I'll do a review + test pass tomorrow 👍

**Comment by zamar-roura** [comment]
Hi @villebro, I'm at your full disposal to improve the merge as much as needed, it's my first merge in superset, if anything is not as it should be tell me and I'll put my full effort into it.

**Comment by villebro** [comment]
@zamar-roura I haven't forgotten about this, I've just been insanely busy. But I'll do my absolute best to review this today!

**Comment by zamar-roura** [comment]
@villebro I totally understand, thank you so much for the quick reply!

**Comment by villebro** [comment]
Restarted CI, the cypress workflow seems to be having issues

File: superset/config.py [hunk 1]
```diff
@@ -873,7 +873,8 @@ class CeleryConfig:  # pylint: disable=too-few-public-methods
 # query costs before they run. These EXPLAIN queries should have a small
 # timeout.
 SQLLAB_QUERY_COST_ESTIMATE_TIMEOUT = int(timedelta(seconds=10).total_seconds())
-# The feature is off by default, and currently only supported in Presto and Postgres.
+# The feature is off by default, and currently only supported in Presto and Postgres,
+# and Bigquery.
```
Comment (villebro) [inline superset/config.py:877]: The current config, docs and the examples in `config.py` are in a pretty horrible state right now. I observed the following:

1. The `ESTIMATE_QUERY_COST` config is in fact a feature flag and needs to be moved to `DEFAULT_FEATURE_FLAGS`: https://github.com/apache/superset/blob/49f1cfc3f99478a3394fccddc50ab9bf7f5c4ae4/superset/config.py#L879 This should be updated
2. This example is broken: https://github.com/apache/superset/blob/49f1cfc3f99478a3394fccddc50ab9bf7f5c4ae4/superset/config.py#L905 . It should in fact be `# QUERY_COST_FORMATTERS_BY_ENGINE = {"postgresql": postgres_query_cost_formatter}`. This should also be updated.
3. There are no docs for this feature. In the deprecated docs there's a SQL Lab section that seems to have been lost over the years, and the content is also incorrect: https://apache-superset.readthedocs.io/en/latest/sqllab.html I don't expect you to add docs for this, but just mentioning this here to call attention to it. FYI @rusackas

**Review by villebro — COMMENTED** [review]
Thanks for the contribution and for fixing this, as this feature appears to be broken right now 🙁 Added a few comments after my first pass review and test.

File: superset/config.py (hunk 1 above)
Comment (rusackas) [reply superset/config.py:877]: Thanks for the callout... putting it on my very long to-do list!

Comment (zamar-roura) [reply superset/config.py]: Fixed the things and added some more comments.

About the docs I would love to fix the content. Only thing is that the superset/blob/master/docs/sqllab.rst file doesnt even exist now.

Comment (villebro) [reply superset/config.py]: @zamar-roura let's leave the docs for a follow-up PR 👍

File: superset/db_engine_specs/bigquery.py [hunk 2]
```diff
@@ -364,6 +365,98 @@ def df_to_sql(
...
+                )
+
+            return {f"{byte_type} Processed": total_bytes_processed}
+        return {}
+
+    @classmethod
+    def query_cost_formatter(
+        cls, raw_cost: List[Dict[str, Any]]
+    ) -> List[Dict[str, str]]:
+        print([{k: str(v) for k, v in row.items()} for row in raw_cost])
```
Comment (villebro) [inline superset/db_engine_specs/bigquery.py:457]: Is this a leftover from debugging?
```suggestion
        print([{k: str(v) for k, v in row.items()} for row in raw_cost])
```

**Review by villebro — COMMENTED** [review]
A few last minute comments/questions

**Comment by zamar-roura** [comment]
Hi @villebro. I don't remember if there was anything else we needed to do. Its a busy end of the year and also full of festivities so if you want we can see what´s needed after all this ends, cheers!

**Comment by zamar-roura** [comment]
@villebro I'll be checking from time to time to resolve merge conflicts

**Review by villebro — APPROVED** [review]
LGTM, thanks for your patience with this PR. Also, thanks for bringing up #18694 which I appear to have dropped the ball on 🙁

**Pull request merged**